pennylane/devices/qubit/apply_operation.py

```diff
@@ -0,0 +1,179 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""appling operation to a statevector."""
+from functools import singledispatch
+from string import ascii_letters as ABC
+
+import pennylane as qml
+
+from pennylane import math
+
+SQRT2INV = 1 / math.sqrt(2)
+
+
+def apply_operation_einsum(op: qml.operation.Operator, state):
+    """Apply ``Operator`` to ``state`` using ``matheinsum``. This is more efficent at lower qubit
+    numbers.
+
+    Args:
+        op (Operator): Operator to apply to the
+        state (array[complex]): input state
+
+    Returns:
+        array[complex]: output_state
+    """
+    mat = op.matrix()
+
+    total_indices = len(state.shape)
+    num_indices = len(op.wires)
+
+    state_indices = ABC[:total_indices]
+    affected_indices = "".join(ABC[i] for i in op.wires)
+
+    new_indices = ABC[total_indices : total_indices + num_indices]
+
+    new_state_indices = state_indices
+    for old, new in zip(affected_indices, new_indices):
+        new_state_indices = new_state_indices.replace(old, new)
+
+    einsum_indices = f"{new_indices}{affected_indices},{state_indices}->{new_state_indices}"
+
+    reshaped_mat = math.reshape(mat, [2] * (num_indices * 2))
+
+    return math.einsum(einsum_indices, reshaped_mat, state)
+
+
+def apply_operation_tensordot(op: qml.operation.Operator, state):
+    """Apply ``Operator`` to ``state`` using ``math.tensordot``. This is more efficent at higher qubit
+    numbers.
+
+    Args:
+        op (Operator): Operator to apply to the
+        state (array[complex]): input state
+
+    Returns:
+        array[complex]: output_state
+    """
+    mat = op.matrix()
+    total_indices = len(state.shape)
+    num_indices = len(op.wires)
+    reshaped_mat = math.reshape(mat, [2] * (num_indices * 2))
+    axes = (tuple(range(num_indices, 2 * num_indices)), tuple(op.wires))
+
+    tdot = math.tensordot(reshaped_mat, state, axes=axes)
+
```

Comment:
It would be worth to readd the commentary on why it is necesary.

Reply:
copied comment over from default qubit.